realm/src/main/java/io/realm/internal/android/AsyncRealmQuery.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2015 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.internal.android;
+
+import android.os.Handler;
+import android.os.Looper;
+import android.os.Message;
+
+import java.util.concurrent.Future;
+
+import io.realm.AsyncRealmQueryResult;
+import io.realm.Realm;
+import io.realm.RealmConfiguration;
+import io.realm.RealmObject;
+import io.realm.RealmQuery;
+import io.realm.RealmResults;
+import io.realm.internal.TableOrView;
+import io.realm.internal.TableView;
+
+import static io.realm.Realm.asyncQueryExecutor;
+
+/**
+ * A helper class to help make handling asynchronous {@link RealmQuery} easier.
+ *
+ * @param <E> type of the object which is to be queried for
+ */
+// Behaviour may change, user should not subclass
+public final class AsyncRealmQuery<E extends RealmObject> {
```

Comment:
This is needed to differ the call to Core until the user specify the last predicate `findAll` `findFirst` etc. This is necessary because the behaviour of the classic `RealmQuery` is to start calling Core as the user chain method calls, as a side effect (the code will still run in the caller Realm/thread) which is not what we're trying to achieve here